File: Data Structures & Algorithms/redundant-connection/submission-0.py
class Solution:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        n = len(edges)
        adj = [[] for _ in range(n + 1)]

        def dfs(node, prev):
            if visit[node]:
                return True

            visit[node] = True
            for nei in adj[node]:
                if nei == prev:
                    continue
                if dfs(nei, node):
                    return True

        for e1, e2 in edges:
            visit = [False] * (n + 1)
            adj[e1].append(e2)
            adj[e2].append(e1)
            
            if dfs(e1, -1):
                return [e1, e2]
        return []

File: Data Structures & Algorithms/cheapest-flight-path/submission-0.java
class Solution {
    public int findCheapestPrice(int n, int[][] flights, int src, int dst, int k) {
        Map<Integer, List<int[]>> adj = new HashMap<>();
        for(int i=0; i<n; ++i) {
            adj.put(i, new ArrayList<>());
        }
        for(int[] flight : flights) {
            adj.get(flight[0]).add(new int[]{flight[2], flight[1]});
        }
        PriorityQueue<int[]> minHeap = new PriorityQueue<>((a, b) -> Integer.compare(a[0], b[0]));
        minHeap.add(new int[]{0, src, 0});
        while(!minHeap.isEmpty()) {
            int[] cur = minHeap.poll();
            int cost = cur[0], node = cur[1];
            // System.out.println(cost + " " + node + " " + cur[2]);
            if(cur[1] == dst && cur[2] - 1 <= k) {
                return cost;
            }
            for(int[] nei : adj.get(node)) {
                int neiK = cur[2];
                minHeap.add(new int[]{cost + nei[0], nei[1], ++neiK});
            }
        }
        return -1;
    }
}
